src/Controls/src/Core/SetterSpecificityList.cs

```diff
@@ -1,127 +1,86 @@
-					[_second.Value.Key] = _second.Value.Value,
-				};
-				// Clear the fields, to reduce duplication in memory
-				_first = null;
-				_second = null;
-			}
-			_values[specificity] = value;
+			var clearedValueIndex = Count - 2;
+			return clearedValueIndex >= 0 ? GetSpecificityAt(clearedValueIndex) : default!;
 		}
-
-		public void Remove(SetterSpecificity specificity)
+		
+		/// <summary>
+		/// Gets the highest specificity
+		/// </summary>
+		/// <returns></returns>
+		public SetterSpecificity GetSpecificity()
 		{
-			_values?.Remove(specificity);
-			if (_first is not null && _first.Value.Key == specificity)
-				_first = null;
-			if (_second is not null && _second.Value.Key == specificity)
-				_second = null;
+			var count = Count;
+			return count == 0 ? default : GetSpecificityAt(count - 1);
 		}
-
-		public KeyValuePair<SetterSpecificity, object> GetSpecificityAndValue()
+		
+		/// <summary>
+		/// Gets the value for the highest specificity
+		/// </summary>
+		/// <returns></returns>
+		public T GetValue()
 		{
-			// Slow path calls SortedList.Last()
-			if (_values is not null)
-				return _values.Last();
-
-			// Fast path accesses _first and _second
-			if (_first is not null && _second is not null)
-			{
-				if (_first.Value.Key.CompareTo(_second.Value.Key) >= 0)
-				{
-					return _first.Value;
-				}
-				else
-				{
-					return _second.Value;
-				}
-			}
-			else if (_first is not null)
-			{
-				return _first.Value;
-			}
-			else if (_second is not null)
-			{
-				return _second.Value;
-			}
-
-			throw new InvalidOperationException("No BindablePropertyContext Value specified!");
+			var count = Count;
+			return count == 0 ? default! : GetSpecificityValueAt(count - 1);
 		}
 
 		/// <summary>
-		/// Called by ClearValueCore, returns what the value would be if cleared
+		/// Gets the highest specificity and value
 		/// </summary>
-		public object? GetClearedValue()
+		/// <returns></returns>
+		public KeyValuePair<SetterSpecificity, T> GetSpecificityAndValue()
 		{
-			if (_values is not null)
-			{
-				return _values.Count >= 2 ? _values[_values.Keys[_values.Count - 2]] : null;
-			}
+			var count = Count;
 
-			// Fast path should return the "lower" value
-			if (_first is not null && _second is not null)
+			if (count == 0)
 			{
-				if (_second.Value.Key.CompareTo(_first.Value.Key) >= 0)
-				{
-					return _first.Value.Value;
-				}
-				else
-				{
-					return _second.Value.Value;
-				}
+				return default;
 			}
-			else if (_first is not null)
-			{
-				return _first.Value.Value;
-			}
-			else if (_second is not null)
-			{
-				return _second.Value.Value;
-			}
-
-			return null;
+			
+			return new KeyValuePair<SetterSpecificity, T>(GetSpecificityAt(count - 1), GetSpecificityValueAt(count - 1));
 		}
+		
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		SetterSpecificity GetSpecificityAt(int index)
+		{
+#if NETSTANDARD
+			return Keys[index];
+#else
+			return GetKeyAtIndex(index);
```

Comment:
Do you have an idea how much time is spent now in `GetKeyAtIndex()`?

I was using `dotnet-trace` on this app: https://github.com/jonathanpeppers/lols

Reply:
The same time required to access a `TKey[]` array index, so it shouldn't be a concern.
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections/src/System/Collections/Generic/SortedList.cs#L561-L566

On the opposite side, in `NETSTANDARD` that has some overhead which we could solve via reflection considering that those private properties cannot change due to serialization matters.
```cs
 public class SortedList<TKey, TValue> :
        IDictionary<TKey, TValue>, IDictionary, IReadOnlyDictionary<TKey, TValue> where TKey : notnull
    {
        private TKey[] keys; // Do not rename (binary serialization)
        private TValue[] values; // Do not rename (binary serialization)
```

Anyway I think that with MAUI the goal is always to use the latest target framework available, so I think the current implementation is acceptable.

Reply:
.NET Standard is not really a supported platform and is _just_ used by the Visual Studio IDE for the analyzers and/or compilers. All other systems should be using the netX.0 TFMs.

Reply:
> Do you have an idea how much time is spent now in GetKeyAtIndex()?

The reason I ask, is because `SortedList` usage shows up quite a bit in `dotnet-trace` for me in the past. Before https://github.com/dotnet/maui/pull/17527, like ~1.4% of time in my [LOLs sample](https://github.com/jonathanpeppers/lols) was spent just iterating a `SortedList`.

![image](https://github.com/dotnet/maui/assets/840039/c1fc3735-cac4-4d05-a229-af64e45c801a)

These were the old `.speedscope` files: 

[net7vnet8rc1.zip](https://github.com/dotnet/maui/files/12874829/net7vnet8rc1.zip)

Reply:
I got rid of 1 and 3 by eliminating Linq.
Number 2 I cannot understand from the screenshot to which call it refers.
7 can be eliminated by giving a initial capacity to the sorted list.

Anyway, there is some space for improvement, I'm working on a benchmark project to get the best result possible.
I'll get back to you in a couple of days. Thanks.

Reply:
The `.zip` file above has files you can open in https://speedscope.app/

When you finish your next iteration, I should have some time to `dotnet-trace` these changes and compare. Thanks!

Reply:
@jonathanpeppers I've finished up doing an ad hoc implementation which brings great results.
You can [look at them here](https://github.com/albyrock87/setter-specificity-list-benchmarks/).

I've updated the PR and also the description above with new benchmark results.

Reply:
Today I've also added on the other repo an implementation based on an unordered linked list.
It looks better from performance side, probably due to the low number of Setters (log(N) vs N is not that relevant with low numbers), but putting it here gave me a worse result in terms of memory usage on the first benchmark (129 vs 121):

|                      Method |     Mean |   Error |  StdDev |    Gen0 |   Gen1 | Allocated |
|---------------------------- |---------:|--------:|--------:|--------:|-------:|----------:|
| SettersFromDifferentSources | 147.9 us | 1.84 us | 1.72 us | 20.9961 | 8.3008 | 129.39 KB |


|           Method |     Mean |    Error |   StdDev |   Gen0 |   Gen1 | Allocated |
|----------------- |---------:|---------:|---------:|-------:|-------:|----------:|
|         BindName | 19.68 us | 0.423 us | 1.248 us | 1.6785 | 1.6479 |  10.31 KB |
|        BindChild | 30.06 us | 0.601 us | 1.255 us | 2.5940 | 2.5635 |  16.02 KB |
| BindChildIndexer | 54.53 us | 1.068 us | 1.663 us | 4.1504 | 4.0894 |  25.55 KB |

What really makes the difference between the two implementations in terms of speed is basically the `Array.Copy` which apparently costs more than iterating on a few items.

Anyway I don't see this huge benefit in going with such implementation which may give worse results when N is bigger.
But I'll leave this choice up to you: for now I've kept the `AdHoc` implementation similar to a `SortedList`.